bridge_td_client_rust/src/lib.rs

```diff
@@ -779,3 +781,141 @@ pub extern fn import_commit(
 }
+
+#[no_mangle]
+pub extern fn import_schema(
+    raw_apikey: *const c_char,
+    raw_endpoint: *const c_char,
+    raw_query_engine: *const c_char,
+    raw_database: *const c_char,
+    raw_server: *const c_char,
+    mut commands: *mut c_void,
+    plappend: extern fn(*mut c_void, usize, &[u8]) -> *mut c_void,
+    debug_log: extern fn(usize, &[u8]),
+    error_log: extern fn(usize, &[u8])) -> *mut c_void {
+
+    let apikey = convert_str_from_raw_str(raw_apikey);
+    let endpoint = convert_str_opt_from_raw_str(raw_endpoint);
+    let query_engine = convert_str_from_raw_str(raw_query_engine);
+    let database = convert_str_from_raw_str(raw_database);
+    let server = convert_str_from_raw_str(raw_server);
+
+    match import_schema_internal(apikey, endpoint, query_engine, database,
+                                 server, debug_log, error_log) {
+        Ok(queries) => {
+            for query in &queries {
+                let query_byte = query.as_bytes();
+                commands = plappend(commands, query_byte.len(), query_byte);
+            }
+            commands
+        },
+        Err(_) => null_mut()
+    }
+}
+
+fn import_schema_internal(
+    apikey: &str,
+    endpoint: Option<&str>,
+    query_engine: &str,
+    database: &str,
+    server: &str,
+    debug_log: extern fn(usize, &[u8]),
+    error_log: extern fn(usize, &[u8])) -> Result<Vec<String>, String> {
+    // TODO better error handling
+    log!(debug_log, "import_schema: entering");
+    log!(debug_log, "import_schema: apikey.len={:?}", apikey.len());
+    log!(debug_log, "import_schema: endpoint={:?}", endpoint);
+    log!(debug_log, "import_schema: query_engine={:?}", query_engine);
+    log!(debug_log, "import_schema: database={:?}", database);
+    log!(debug_log, "import_schema: server={:?}", server);
+
+    let client = create_client(apikey, &endpoint);
+    let query_type = match QueryType::from_str(query_engine) {
+        Ok(query_type) => query_type,
+        Err(err) => {
+            log!(error_log, "import_schema: Invalid query engine. query_engine={:?}. error={:?}", 
+                    query_engine, err);
+            return Err(String::from("invalid query engine"))
+        }
+    };
+    log!(debug_log, "import_schema: query_type={:?}", query_type);
+
+    let tables = match Retry::new(
+        &mut || client.tables(database),
+        &mut |result| test_if_needs_to_retry(result)
+    ).try(RETRY_COUNT).wait(RETRY_FIXED_INTERVAL_MILLI).execute() {
+        Ok(Ok(tables)) => tables,
+        Ok(Err(err)) => {
+            log!(error_log, "import_schema: Failed to import schema. {:?}", err);
+            return Err(String::from("td API error"))
+        },
+        Err(err) => {
+            log!(error_log, "import_schema: Failed to import schema. {:?}", err);
+            return Err(String::from("td API error"))
+        }
+    };
+
+    let mut commands = Vec::new();
+    'outer: for tab in &tables {
```

Comment:
[minor] This kind complicated labeled loop might be replaced with simple recursive function calls.